src/API.jl

```diff
@@ -894,35 +894,42 @@ end
 
 precompile() = precompile(Context())
 function precompile(ctx::Context)
+    
+    function is_stale(paths, sourcepath)
+        for path_to_try in paths::Vector{String}
+            staledeps = Base.stale_cachefile(sourcepath, path_to_try, Base.TOMLCache())
+            staledeps === true && continue
+            return false
+        end
+        return true
+    end
+    
+    # stdlibs that aren't loaded might need precompiling if not part of sysimage
+    function is_stdlib_and_loaded(pkg::Base.PkgId)
+        return Pkg.Operations.is_stdlib(pkg.uuid) && Base.root_module_exists(pkg)
+    end
+    
     printpkgstyle(ctx, :Precompiling, "project...")
     
     num_tasks = parse(Int, get(ENV, "JULIA_NUM_PRECOMPILE_TASKS", string(Sys.CPU_THREADS + 1)))
     parallel_limiter = Base.Semaphore(num_tasks)
     
+    toplevel_pkgids = [Base.PkgId(uuid, name) for (name, uuid) in ctx.env.project.deps if !is_stdlib_and_loaded(Base.PkgId(uuid, name))]
+    
     man = Pkg.Types.read_manifest(ctx.env.manifest_file)
-    pkgids = [Base.PkgId(first(dep), last(dep).name) for dep in man if !Pkg.Operations.is_stdlib(first(dep))]
-    pkg_dep_uuid_lists = [collect(values(last(dep).deps)) for dep in man if !Pkg.Operations.is_stdlib(first(dep))]
-    filter!.(!is_stdlib, pkg_dep_uuid_lists)
+    pkgids = [Base.PkgId(first(dep), last(dep).name) for dep in man if !is_stdlib_and_loaded(Base.PkgId(first(dep), last(dep).name))]
+    pkg_dep_lists = [[Base.PkgId(last(x), first(x)) for x in last(dep).deps if !is_stdlib_and_loaded(Base.PkgId(last(x), first(x)))] for dep in man if !is_stdlib_and_loaded(Base.PkgId(first(dep), last(dep).name))]
```

Comment:
This has got a bit too long.. is there a more concise way to generate a vector of vector of deps of deps as PkgIds?

Reply:
Please use 92 width limit. FYI, JuliaFormatter.jl would format this to

```julia
    pkg_dep_lists = [
        [
            Base.PkgId(last(x), first(x))
            for
            x in last(dep).deps if !is_stdlib_and_loaded(Base.PkgId(last(x), first(x)))
        ]
        for dep in man if !is_stdlib_and_loaded(Base.PkgId(first(dep), last(dep).name))
    ]
```

which is reasonable, I think.

Also, I just noticed that it's crucial that the orders/indices of `pkgids` and `pkg_dep_lists` are consistent.  I think it'd be better to merge them in a single data structure like a vector-of-pairs or a dictionary (so that it's impossible to break such "consistency" by construction):

```julia
    depsmap = Dict(
        Base.PkgId(first(dep), last(dep).name) => [
            Base.PkgId(last(x), first(x))
            for
            x in last(dep).deps if !is_stdlib_and_loaded(Base.PkgId(last(x), first(x)))
        ]
        for dep in man if !is_stdlib_and_loaded(Base.PkgId(first(dep), last(dep).name))
    )
```

To me it's a bit hard to read because of repeated `Base.PkgId(...)`. I'd probably go with something like

```julia
    deps_pair_or_nothing = Iterators.map(man) do dep
        pkg = Base.PkgId(first(dep), last(dep).name)
        is_stdlib_and_loaded(pkg) && return nothing
        deps = (Base.PkgId(last(x), first(x)) for x in last(dep).deps)
        return pkg => filter(!is_stdlib_and_loaded, deps)
    end
    depsmap = Dict(Iterators.filter(!isnothing, deps_pair_or_nothing))
```

Reply:
Thanks. That last form hits this, which I'm not sure how to handle
```
ERROR: MethodError: no method matching filter(::Base.var"#70#71"{Pkg.API.var"#is_stdlib_and_loaded#210"}, ::Base.Generator{Dict{String, Base.UUID}, Pkg.API.var"#204#220"})
Closest candidates are:
  filter(::Any, ::Tuple{Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Vararg{Any, N}} where N) at tuple.jl:311
  filter(::Any, ::Tuple) at tuple.jl:308
  filter(::Any, ::Array{T, N}) where {T, N} at array.jl:2469
```

Reply:
OK, so `filter(f, generator)` does not work... I think this should be fixed in `Base`. Meanwhile, I'd try something like this

```julia
        deps = [Base.PkgId(last(x), first(x)) for x in last(dep).deps]
        return pkg => filter!(!is_stdlib_and_loaded, deps)
```

Reply:
Thanks. Fixed now